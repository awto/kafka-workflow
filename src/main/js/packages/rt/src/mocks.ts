/**
 * lightweight workflow's testing
 *
 * the workflow functions are usual JS async functions, and can be debugged and tested as any other JS function
 */
export const config = {
  outputTopics: new Set(["workflow-result", "workflow-error"]),
  resultTopic: "workflow-result",
  errorTopic: "workflow-error"
};

export let threadId = "testThread";
export let stepId = "testStep";

export const __topicHandlers = new Map<
  string,
  (msg: string, dest: string, topic: string) => unknown
>();

interface Output {
  msg: string;
  dest: string;
  topic: string;
}

export const __trace: any[] = [];
export const __suspended = new Map<string, Suspended>();
export let __lastTopicUpdate: {[topicName:string]:{k:string,v:string}} = {};

export async function __resume(id: string, value: unknown) {
  __trace.push({resume:id, value});
  await new Promise(i => setTimeout(i, 0))
  const suspended = __suspended.get(id);
  if (!suspended) throw new Error(`nothing is suspended with id ${id}`);
  __suspended.delete(id);
  for (const i of suspended.onfulfilled) i(value);
  await new Promise(i => setTimeout(i, 0))
}

export async function __reject(id: string, reason: unknown) {
  await new Promise(i => setTimeout(i, 10))
  __trace.push({reject:id, reason});
  const suspended = __suspended.get(id);
  if (!suspended) throw new Error(`nothing is suspended with id ${id}`);
  __suspended.delete(id);
  for (const i of suspended.onrejected) i(reason);
  // suspended.onrejected && suspended.onrejected(reason);
  await new Promise(i => setTimeout(i, 10))
}

/** resets state in tests setup */
export function __reset() {
  __lastTopicUpdate = {}
  __trace.length = 0;
  Suspended.count = 0;
  __topicHandlers.clear();
}

export class Suspended {
  static count = 0;
  onfulfilled: ((arg: unknown) => unknown)[] = [];
  onrejected: ((arg: unknown) => unknown)[] = [];
  constructor(public id: string = `${stepId}:${++Suspended.count}`) {
    __suspended.set(id, this);
  }
  then(onff: (arg: unknown) => unknown, onrj?: (arg: unknown) => unknown) {
    this.onfulfilled.push(onff);
    if (onrj) this.onrejected.push(onrj);
  }
}

export function suspend(id?: string): any {
  return new Suspended(id);
}

export function output(msg: string, topic: string, dest = threadId) {
  const item = { dest, msg, topic };
  __trace.push(item);
  const handler = __topicHandlers.get(topic);
  __lastTopicUpdate[topic] = {k:dest, v:msg};
  if (handler) handler(msg, dest, topic);
}

export function outputJSON(msg: unknown, topic: string, dest = threadId) {
  return output(JSON.stringify(msg), topic, dest);
}

export const any = (arr:any[]) => Promise.race(arr);
export const all = (arr:any[]) => Promise.all(arr);
export const anyWithCancel = any;
export const allWithCancel = all;
