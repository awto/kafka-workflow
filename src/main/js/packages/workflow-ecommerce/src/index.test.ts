import * as W from "@effectful/kafka-workflow/lib/mocks";
import * as Index from "./index";

afterEach(() => {
  W.__reset();
});

test("cart operations", async function () {
  const res = Index.main({ abandonedCartTimeoutMS: 50 });
  await W.__resume("main", { type: "checkout" });
  await W.__resume("main", {
    type: "updateEmail",
    email: "someone@example.com"
  });
  await W.__resume("main", { type: "checkout" });
  await W.__resume("main", {
    type: "addToCart",
    item: { quantity: 10, productId: "teapot" }
  });
  await W.__resume("main", {
    type: "addToCart",
    item: { quantity: 11, productId: "sigar" }
  });
  await W.__resume("main", {
    type: "addToCart",
    item: { quantity: 20, productId: "teapot" }
  });
  await W.__resume("main", {
    type: "updateEmail",
    email: "someone.else@example.com"
  });
  await W.__resume("main", {
    type: "removeFromCart",
    item: { quantity: 11, productId: "sigar" }
  });
  await W.__resume("main", {
    type: "addToCart",
    item: { quantity: 2, productId: "sugar" }
  });
  await W.__resume("main", { type: "checkout" });
  const item = await res;
  expect(item).toEqual({
    email: "someone.else@example.com",
    items: [
      {
        productId: "teapot",
        quantity: 30
      },
      {
        productId: "sugar",
        quantity: 2
      }
    ]
  });
  expect(W.__trace).toMatchSnapshot();
  expect(W.__suspended.has("main")).toBe(false);
});

function setUpScheduler(data: { ref: string }) {
  W.__topicHandlers.set(
    "workflow-scheduler",
    function (_msg: string, dest: string) {
      const [thread, ref] = dest.split("|");
      expect(thread).toBe(W.threadId);
      data.ref = JSON.parse(ref).ref;
    }
  );
}

function curRef() {
  const [thread, ref] = W.__lastTopicUpdate["workflow-scheduler"].k.split("|");
  expect(thread).toBe(W.threadId);
  return JSON.parse(ref).ref;
}

test("abondoned and resumed", async function () {
  const res = Index.main({ abandonedCartTimeoutMS: 50 });
  await W.__resume("main", {
    type: "updateEmail",
    email: "someone@example.com"
  });
  await W.__resume(curRef(), undefined);
  await W.__resume("main", {
    type: "addToCart",
    item: { quantity: 2, productId: "sugar" }
  });
  await W.__resume("main", { type: "checkout" });
  const item = await res;
  expect(item).toEqual({
    email: "someone@example.com",
    items: [
      {
        productId: "sugar",
        quantity: 2
      }
    ]
  });
  expect(W.__trace).toMatchSnapshot();
  expect(W.__suspended.has("main")).toBe(false);
});

test("abondoned forever", async function () {
  const res = Index.main({ abandonedCartTimeoutMS: 50 });
  await W.__resume("main", {
    type: "updateEmail",
    email: "someone@example.com"
  });
  await W.__resume(curRef(), undefined);
  await W.__resume(curRef(), undefined);
  const item = await res;
  expect(item).toBe("abondoned");
  expect(W.__trace).toMatchSnapshot();
});
