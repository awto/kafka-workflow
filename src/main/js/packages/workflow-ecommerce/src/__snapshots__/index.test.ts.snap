// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`abondoned and resumed 1`] = `
Array [
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:1\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "email": "someone@example.com",
      "type": "updateEmail",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:2\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "testStep:2",
    "value": undefined,
  },
  Object {
    "dest": "testThread",
    "msg": "someone@example.com",
    "topic": "ecommerce-reminder",
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:3\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "item": Object {
        "productId": "sugar",
        "quantity": 2,
      },
      "type": "addToCart",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:4\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "type": "checkout",
    },
  },
]
`;

exports[`abondoned forever 1`] = `
Array [
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:1\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "email": "someone@example.com",
      "type": "updateEmail",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:2\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "testStep:2",
    "value": undefined,
  },
  Object {
    "dest": "testThread",
    "msg": "someone@example.com",
    "topic": "ecommerce-reminder",
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:3\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "testStep:3",
    "value": undefined,
  },
]
`;

exports[`cart operations 1`] = `
Array [
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:1\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "type": "checkout",
    },
  },
  Object {
    "dest": "testThread",
    "msg": "Must have items to check out!",
    "topic": "checkoutError",
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:2\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "email": "someone@example.com",
      "type": "updateEmail",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:3\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "type": "checkout",
    },
  },
  Object {
    "dest": "testThread",
    "msg": "Must have items to check out!",
    "topic": "checkoutError",
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:4\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "item": Object {
        "productId": "teapot",
        "quantity": 30,
      },
      "type": "addToCart",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:5\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "item": Object {
        "productId": "sigar",
        "quantity": 0,
      },
      "type": "addToCart",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:6\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "item": Object {
        "productId": "teapot",
        "quantity": 20,
      },
      "type": "addToCart",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:7\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "email": "someone.else@example.com",
      "type": "updateEmail",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:8\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "item": Object {
        "productId": "sigar",
        "quantity": 11,
      },
      "type": "removeFromCart",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:9\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "item": Object {
        "productId": "sugar",
        "quantity": 2,
      },
      "type": "addToCart",
    },
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:10\\"}",
    "msg": "50",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "main",
    "value": Object {
      "type": "checkout",
    },
  },
]
`;
